sources/constraint-declaration-validation.asciidoc

```diff
@@ -1026,6 +1100,289 @@ public class SimpleOrderService extends OrderService {
 
 The return value constraints in [classname]`DefaultOrderService` are legal, as they strengthen the postconditions of the [methodname]`placeOrder()` method as constituted by the super class [classname]`OrderService` but don't weaken them.
 
+[[constraintdeclarationvalidationprocess-containerelementconstraints]]
+=== Container element constraints
+
+Constraints can be applied to the elements of generic containers, e.g. `List`, `Map` or `Optional`.
+This is done by putting constraint annotations to the type arguments of such containers.
+Container element constraints can be applied to containers that are
+
+* fields,
+* properties,
+* method or constructor parameters or
+* method return values.
+
+.Container element constraints
+====
+[source, JAVA]
+----
+private List<@Email String> emails;
+
+public Optional<@Email String> getEmail() {
+    [...]
+}
+
+public Map<@NotNull String, @ValidAddress Address> getAddressesByType() {
+    [...]
+}
+
+public List<@NotBlank String> getMatchingRecords(List<@NotNull @Size(max=20) String> searchTerms) {
+    [...]
+}
+----
+====
+
+When a field, property, executable parameter or method return value is validated,
+any container element constraints of that element will be validated alongside any other constraints hosted by that element.
+For container element constraints, the same rules for validation groups and group sequences apply as for any other constraint on the same element.
+
+When a container element constraint is detected, the validation engine invokes a value extractor (see <<constraintdeclarationvalidationprocess-valueextractordefinitions>>)
+which retrieves the value(s) from the container so they can be validated.
+This may be a single value - e.g. in the case of `Optional` which wraps exactly one value if the `Optional` is not empty -
+or multiple values in the case of collection types.
+
+Container element constraints can be applied to nested container types:
+
+.Container element constraints on nested containers
+====
+[source, JAVA]
+----
+private Map<String, @NotEmpty List<@ValidAddress Address>> addressesByType;
+----
+====
+
+In such case multiple value extractors will be invoked.
+In the example above,
+
+* the extractor for `Map` values will be invoked to obtain all map values (lists of `Address`)
+* for each extracted list of addresses, the `@NotEmpty` constraint will be validated and the extractor for `List` elements will be invoked, providing the `Address` objects from each list in the map
+* the `@ValidAddress` constraint will be applied to all elements of all lists stored in the map
+
+Container element constraints must not be put to wildcard type arguments (with or without bounds), i.e. the following usage is disallowed:
```

Comment:
I changed it to "unsupported", although the only thing theoretically making sense is an extends bound (`? extends String`), as otherwise no constraint validator can be resolved.